### PR TITLE
Fix some data races by using volatile when necessary

### PR DESCRIPTION
**Pull request by fabbing**
In our (with @OlivierNicole) ongoing work to integrate TSan to OCaml, TSan has reported some more data races in the runtime, which we have added to #11040.

This PR proposes to fix some of the low-hanging fruits by using relaxed atomic operations where necessary for correctness.

* In `do_some_marking`, when scanning block fields, a data race can occur if another domain modifies a field with `caml_modify`. The read operation should be a relaxed atomic load.
* In `mark_stack_push_range` reading the block header or block fields can race with another domain performing `caml_darken`. Both domains attempting to darken the same block is fine but the read operation should also be relaxed atomic load.

### CODE REVIEW

File: runtime/caml/platform.h [hunk 1]
```diff
@@ -61,6 +61,11 @@ Caml_inline uintnat atomic_load_relaxed(atomic_uintnat* p)
   return atomic_load_explicit(p, memory_order_relaxed);
 }
 
+Caml_inline value atomic_load_relaxed_val(value* p)
+{
+  return atomic_load_explicit(Op_atomic_val(p), memory_order_relaxed);
+}
```
Comment (gadmm) [inline runtime/caml/platform.h]: Instead of duplicating the function to work with different signatures, would it help to make it a macro so that it is polymorphic? (e.g. as in https://github.com/gadmm/ocaml/compare/custom_thread_safety2..custom_thread_safety)
Would that help with writing the code differently?

File: runtime/caml/mlvalues.h [hunk 2]
```diff
@@ -226,6 +230,10 @@ where 0 <= R <= 31 is HEADER_RESERVED_BITS, set with the
 /* Fields are numbered from 0. */
 #define Field(x, i) (((volatile value *)(x)) [i]) /* Also an l-value. */
 
+#define Field_atomic(x, i) \
+  atomic_load_explicit(Op_atomic_val((volatile value *)(x) + i), \
+                       memory_order_relaxed)
```
Comment (gadmm) [inline runtime/caml/mlvalues.h]: If this is a problem for tsan then this will also likely affect FFI code (legacy and new) written by users. Does this mean that adding `volatile` to `Field` was not enough of a fix in OCaml 5 because this will create false positives for tsan?
The original idea was that a volatile read would behave closely-enough to a relaxed read: is there by chance an option to convince tsan of that?



Also, if this is really needed, then some time ago it was discussed that it would be beneficial to split the Field macro in two (one l-value and one r-value). This suggest to simply call this macro Read_field or Field_load or something like that.

Comment (OlivierNicole) [reply runtime/caml/mlvalues.h]: Actually, I think the `volatile` is not really needed here as we only need to tell the compiler that the field read is atomic. TSan does not see that the type is volatile so it won’t make a difference on its side.

Yes, maybe `Field_atomic_load` would be better to make explicit that it is never an l-value.

Comment (gadmm) [reply runtime/caml/mlvalues.h]: I failed to get my point across. What I am worrying about is that if you need to replace `Field` with `Field_atomic` to avoid a warning from tsan, then maybe one also needs to do the replacement in all of legacy and new C FFI code written by users. Am I missing something?

Comment (gasche) [reply runtime/caml/mlvalues.h]: There seems to be support for volatile instrumentation in both gcc and clang TSan implementations using the `tsan-distinguish-volatile` option. See [there](https://lore.kernel.org/all/20200521142047.169334-4-elver@google.com/) for how this is used in the Linux kernel, which also uses volatile as part of its memory model (this was the inspiration for @gadmm's suggestion to use volatile in the OCaml runtime I believe).

Comment (gasche) [reply runtime/caml/mlvalues.h]: In other words: you may not need to change anything here if you use the same approach as the Linux kernel of generating versions of volatile reads/writes that explicitly silence TSan.

Comment (gadmm) [reply runtime/caml/mlvalues.h]: Note that you will still have false positives for dependency ordering, see e.g. https://github.com/google/sanitizers/issues/882

Comment (OlivierNicole) [reply runtime/caml/mlvalues.h]: I see, enabling this option does sound useful, if non-trivial (apparently one must implement the instrumentation functions for `volatile` oneself, and the behavior differs quite a lot between GCC and Clang). We’ll see whether it allows to remove warnings on the runtime.

However, the use of `volatile` is still an ersatz of real atomic accesses intended for legacy code. I think that we should strive to use atomic accesses in the runtime where necessary, provided that it does not require too many changes.

Comment (gasche) [reply runtime/caml/mlvalues.h]: I don't agree with your reasoning. It took a lot of discussion to decide on which style we wanted for our C code, and https://github.com/ocaml/ocaml/issues/10992 proposed to use `volatile` to enforce our memory model, for performance reasons. We collectively agreed on that.

Now you propose to deviate from this specific in a few places to silence TSan warnings. I think this is the wrong approach. My thinking:
1. Either we believe that the "volatile" model that we agreed on is the right design, and we should use tools that understand it. (If the Linux kernel uses the same model, then presumably those tools exist *or* there are other people interesting in making them exist.)
2. Or we believe that being able to use TSan unmodified is a strong argument in favor of moving to a different model (using 2A instead of 2B in the classification of @gadmm in https://github.com/ocaml/ocaml/issues/10992#issuecomment-1104033399), and we should discuss moving to a new model *in general*, instead of just in a couple places that currently raise warnings.

Comment (OlivierNicole) [reply runtime/caml/mlvalues.h]: Sure, going with approach 1 is fine by me if this is the conclusion that was reached.

> Now you propose to deviate from this specific in a few places to silence TSan warnings. I think this is the wrong approach.

I didn’t mean to suggest that. TSan is the tool that allowed us to find these potential races (one of which is actually not a race, it appears, thanks to `volatile`), but the object of this PR is to make sure the runtime is correct, not to content TSan. Thanks for the opportunity to clarify that.

Comment (gadmm) [reply runtime/caml/mlvalues.h]: > It took a lot of discussion to decide on which style we wanted for our C code, and https://github.com/ocaml/ocaml/issues/10992 proposed to use volatile to enforce our memory model, for performance reasons.

To clarify, using `volatile` was more of a stop-gap measure. #10992 is still an open issue with two proposed paths, with little to no discussion happening. Since then, in private conversations with one memory-model expert we agreed that one solution does not work, whereas with another memory-model expert we agreed that the other solution would not work. (I prefer to clarify that I do not consider myself an expert on memory models and more of an observer as far as #10992 is concerned.)

In particular if you instrument volatile accesses as relaxed reads for TSan, you will still have a lot of false positives (for the publication stuff), whereas if you instrument it as an acquire read for TSan, then I suspect you might have a lot of false negatives. Either way it is already good to have experience from trying to use TSan as this might affect how we think about the memory-model problem.

**Comment by fabbing** [comment]
Thanks for the link for to #10992, it contains valuable information on the implications of adding such macro.

We've reverted the changes that add atomic read operation to `Tag_*` and `Field` and stick with `volatile` for now.
The last commit adds a `volatile` qualifier to the `Tag_*` macros, which should also be close enough to a relaxed atomic read.
We've kept the `atomic_load_relaxed_val` in `do_some_marking` to show our clear intention to have a relax atomic read here, without the risk of performance impact if we made the `mark_entry` `start` and `end` members to be `volatile value*`.

Using `tsan-distinguish-volatile` with our TSan integration proved to be more difficult than expected.
While `gcc`/`clang` replace `atomic` read/write operations with TSan calls that update TSan's internal state and perform the actual memory operation itself, `volatile` read/write operations are decorated with a TSan call and then the actual operation is performed.

- An example using an atomic read operation
```c
_Atomic uint64_t a = 41;

__attribute__((noinline))
uint64_t an_atomic_read(void)
{
  return a;
}
```
Reading `a` is replaced with a call to `__tsan_atomic64_load` that notifies TSan of the read operation and returns the actual value.

```
__attribute__((noinline))
uint64_t an_atomic_read(void)
{
    117a:       53                      push   %rbx
    117b:       48 8b 7c 24 08          mov    0x8(%rsp),%rdi
    1180:       e8 bb fe ff ff          call   1040 <__tsan_func_entry@plt>
  return a;
    1185:       be 05 00 00 00          mov    $0x5,%esi
    118a:       48 8d 3d af 2e 00 00    lea    0x2eaf(%rip),%rdi        # 4040 <a>
    1191:       e8 ba fe ff ff          call   1050 <__tsan_atomic64_load@plt>
    1196:       48 89 c3                mov    %rax,%rbx
    1199:       e8 c2 fe ff ff          call   1060 <__tsan_func_exit@plt>
}
    119e:       48 89 d8                mov    %rbx,%rax
    11a1:       5b                      pop    %rbx
    11a2:       c3                      ret

```

- An example using a volatile read operation
```c
uint64_t v = 42;

__attribute__((noinline))
uint64_t a_volatile_read(void)
{
  return *(volatile uint64_t*)&v;
}
```
When reading `v` a call to `__tsan_volatile_read8` (at `11b5`) precedes the actual memory read operation (at `11ba`).
```
__attribute__((noinline))
uint64_t a_volatile_read(void)
{
    11a3:       53                      push   %rbx
    11a4:       48 8b 7c 24 08          mov    0x8(%rsp),%rdi
    11a9:       e8 92 fe ff ff          call   1040 <__tsan_func_entry@plt>
  return *(volatile uint64_t*)&v;
    11ae:       48 8d 3d 83 2e 00 00    lea    0x2e83(%rip),%rdi        # 4038 <v>
    11b5:       e8 bf ff ff ff          call   1179 <__tsan_volatile_read8>
    11ba:       48 8b 1d 77 2e 00 00    mov    0x2e77(%rip),%rbx        # 4038 <v>
    11c1:       e8 9a fe ff ff          call   1060 <__tsan_func_exit@plt>
}
    11c6:       48 89 d8                mov    %rbx,%rax
    11c9:       5b                      pop    %rbx
    11ca:       c3                      ret
```
We also need to provide the actual implementation of `__tsan_volatile_read8`.
Our attempt included calling TSan's own `__tsan_atomic64_load` to notify of an atomic relaxed read, but the generated code is then followed by the actual memory read operation, which can be an issue for performance and correctness.

**Comment by gasche** [comment]
The linux kernel provides its own definitions of those volatile access operations (see [there](https://lore.kernel.org/all/20200521142047.169334-4-elver@google.com/)), could you reuse them or at least take inspiration from them?

**Comment by OlivierNicole** [comment]
As mentioned in the commit message of https://github.com/llvm/llvm-project/commit/5a2c31116f412c3b6888be361137efd705e05814, Linux actually uses KCSAN, a runtime that uses TSAN instrumentation but
otherwise implements a very different approach to race detection. For this reason it is not at all obvious how to implement `__tsan_volatile_{read,write}*` in a way that would allow for correct race detection with TSan.

**Comment by gasche** [comment]
Now I get what you are saying, I think: the TSAN interface is weird and the `_volatile_` stuff is hard to use correctly. (Maybe no one has tried to use it for this yet?)

The implementation of the `atomic*_load` functions seem to be this C++ template:

https://github.com/llvm/llvm-project/blob/b4257d3bf58c51ff437c99c7b160023f5fffac78/compiler-rt/lib/tsan/rtl/tsan_interface_atomic.cpp#L223-L247

It looks like there is a `MemoryAccess` function that is used to notify TSan of (relaxed?) memory accesses, and is probably what we want to call from the `_volatile_` callbacks?

**Comment by gasche** [comment]
Actually I have a very naive question, sorry for the trouble: if the access inserted by the instrumentation a "raw" volatile access, then what else do you want your instrumentation operation to do? Would it be correct to just do nothing at all?

I have no idea how TSan works, but I would guess that a synchronization sanitizer wants to instruments non-atomic accesses to emit alerts about some of them, and also to instrument ordered atomic accesses to store some ordering information (to detect that some later non-atomic accesses are safe). But for relaxed atomics, does it need to do anything, can it not just ignore them? (I'm assuming that we want volatile to behave similarly to relaxed atomics.)

**Comment by OlivierNicole** [comment]
That is a very relevant suggestion, I think, and we had the same intuition. The only problem is that the instrumentation of a memory access in TSan performs two things:
- checking for data races with the memory operation being signaled, and
- recording the ordering constraints created by that operation in TSan’s shadow state.

As you point out, a relaxed atomic access (and we assume of operations on `volatile`s) does not incur any ordering constraints. However, not doing anything at all would prevent TSan from detecting data races with the `volatile` access—which can happen, e.g. between a volatile read and a plain store.

However, maybe we can make this work for volatile reads by calling `__tsan_atomicN_load` and ignoring its value. What you suggest about calling `MemoryAccess`, but that function is internal to TSan, and I’m not sure what that entails in terms of possibility of calling it from the outside (C++ symbol mangling comes to mind).

I’m very grateful for this possibility of improving TSan support and I suggest we pursue the discussion in a dedicated issue: https://github.com/ocaml-multicore/ocaml-tsan/issues/25. This way we can keep this thread focused on the proposed changes, independently from TSan.

File: runtime/caml/platform.h (hunk 1 above)
Comment (OlivierNicole) [reply runtime/caml/platform.h]: I suppose it would make some sense, but as it is a change with consequences outside of the scope of this PR, I would be in favor of having a dedicated discussion for it.

Regarding this PR, if the function duplication is a problem, maybe @fabbing could use an atomic cast at the call site, as has been done previously to avoid races, e.g., in https://github.com/ocaml/ocaml/pull/11512.

Comment (gadmm) [reply runtime/caml/platform.h]: If I understand correctly, the race with `*me.start` is because the value resides on the OCaml heap, and thus can be mutated in parallel by OCaml code. 1) Good catch, 2) In this case, I think the way it is done elsewhere is that the type of pointers to values that reside in the OCaml heap is never `value *`, but `volatile value *` or even `volatile const value *`.

Assume we change the types of `mark_entry` elements to `volatile const value *`. This fixes the race. Is this still a problem for OCaml-TSan, or for using TSan to debug the runtime? Because I suspect this is far from the only place where the issue will arise.

- If OCaml-TSan: can we simply assume the GC is correct and exclude this function from TSan instrumentation? (after fixing it with volatile)
- If TSan used to debug the runtime: is it not a bit hopeless to try to get rid of reported races on volatile, given how much code is written with volatile?

Comment (gadmm) [reply runtime/caml/platform.h]: As for what is proposed in this PR (introducing a function that does a cast to `atomic_value *`).
- First the argument should not be a `value *`. Instead, it is better if the mark_entry element itself is an `atomic_value *`. (Replacing `me.start = Op_val(block);` with `me.start = Op_atomic_val(block);`)
- Once the argument is correctly an `atomic_value *`, it is better to make the atomic_* function polymorphic as above (but this can be done in a separate PR).
- I am mainly concerned that this path taken (use atomic instead of volatile) might take us too far, in the sense that it is not possible to go the full length of converting all the C code to use atomic instead of volatile in this style.

Comment (gadmm) [reply runtime/caml/platform.h]: I suggest making a change to `mark_entry` according to what is more appropriate to your needs (`value *` -> `volatile const *` or `value *` -> `atomic_value *`). For the rest, which is about the more general issue of TSan for the runtime, I think we greatly need to understand the more general situation.

Comment (gadmm) [reply runtime/caml/platform.h]: E.g. if we make it atomic here, we would like to apply this style consistently in the runtime; at the same time we cannot change the `Field` macro. Do you see a clear boundary between what should be made atomic, and what must remain volatile, based on TSan needs? For instance, we could say that only the C code which is part of the mutator (parts of the C runtime called from OCaml) should be allowed to use volatile. Would it improve the situation?

Comment (gasche) [reply runtime/caml/platform.h]: My take:
1. The TSan implmentation in either GCC and Clang is experimental, poorly documented and difficult to use; it works well for simple C programs following the C11 memory model to the letter, and poorly outside this subset. The implementation is full of odd special cases for other languages than C.
2. While the pure-C parts of the C runtime use the C11 memory model, the FFI parts of the runtime (the parts that interact with the OCaml heap) use a volatile-based model that @gadmm suggested and corresponds to some parts of the Linux Kernel approach.
3. TSan does not currently support volatile models well enough.
4. In my opinion, a good solution is to improve TSan -- which is a service not just to OCaml but to a much wider community of programmers, including the Linux kernel.

My understanding is that @OlivierNicole has a slightly different view of things (maybe motivated by the idea of using the tools as they are and not broadening the scope of his work too much in the direction of "fix both GCC and Clang"), where he thinks of the `volatile` use in the OCaml runtime as a "deprecated" pattern -- despite the fact that there we offer no support for anything else than this approach right now -- so we could look the other way and hope to eventually convert the runtime to be fully C11-memory-model based, including the cross-language parts.

Comment (OlivierNicole) [reply runtime/caml/platform.h]: @gadmm:

> Assume we change the types of mark_entry elements to volatile const value *. This fixes the race. Is this still a problem for OCaml-TSan, or for using TSan to debug the runtime? Because I suspect this is far from the only place where the issue will arise.

Currently it is still a problem for ocaml-tsan, but we are discussing a fix at https://github.com/ocaml-multicore/ocaml-tsan/issues/25.

> If OCaml-TSan: can we simply assume the GC is correct and exclude this function from TSan instrumentation? (after fixing it with volatile)

Yes, that’s what ocaml-tsan does at the moment.

> If TSan used to debug the runtime: is it not a bit hopeless to try to get rid of reported races on volatile, given how much code is written with volatile?

I wouldn’t say hopeless, because either the aforementioned fix will work, or we can still look at each race report in #11040 individually and close the cases that are solved by `volatile`. As there is a relatively small number of them, this remains manageable.

> E.g. if we make it atomic here, we would like to apply this style consistently in the runtime; at the same time we cannot change the Field macro. Do you see a clear boundary between what should be made atomic, and what must remain volatile, based on TSan needs? For instance, we could say that only the C code which is part of the mutator (parts of the C runtime called from OCaml) should be allowed to use volatile. Would it improve the situation?

It is hard for me to have an opinion on this, but @maranget in https://github.com/ocaml/ocaml/issues/10992#issuecomment-1152124829 advocates for using `volatile` everywhere for performance reasons.

@gasche:

> In my opinion, a good solution is to improve TSan -- which is a service not just to OCaml but to a much wider community of programmers, including the Linux kernel.

I agree with this, and we can propose changes in that direction to the TSan team. There is no guarantee that there will be a result anytime soon, however.

> My understanding is that @OlivierNicole has a slightly different view of things (maybe motivated by the idea of using the tools as they are and not broadening the scope of his work too much in the direction of "fix both GCC and Clang"), where he thinks of the volatile use in the OCaml runtime as a "deprecated" pattern -- despite the fact that there we offer no support for anything else than this approach right now -- so we could look the other way and hope to eventually convert the runtime to be fully C11-memory-model based, including the cross-language parts.

I indeed used to think of it as deprecated, but if the agreement is that `volatile` is preferrable (e.g. for performance), I’m happy to implement this choice (although I can’t speak for @fabbing), regardless of TSan support.

File: runtime/caml/mlvalues.h (hunk 2 above)
Comment (OlivierNicole) [reply runtime/caml/mlvalues.h]: > In particular if you instrument volatile accesses as relaxed reads for TSan, you will still have a lot of false positives (for the publication stuff), […]

Regarding the specific point of publication safety, it should no longer trigger false positives, as we no longer instrument initializing writes. ([Justification of why and explanations here.](https://github.com/ocaml-multicore/ocaml-tsan/pull/22#issuecomment-1372229946))

Comment (gadmm) [reply runtime/caml/mlvalues.h]: @OlivierNicole Regarding initializing writes, did you manage to not instrument the ones done from C code?

File: runtime/caml/mlvalues.h [hunk 3]
```diff
@@ -195,14 +195,14 @@ where 0 <= R <= 31 is HEADER_RESERVED_BITS, set with the
 #define Reserved_val(val) (Reserved_hd (Hd_val (val)))
 
 #ifdef ARCH_BIG_ENDIAN
-#define Tag_val(val) (((unsigned char *) (val)) [-1])
+#define Tag_val(val) (((volatile unsigned char *) (val)) [-1])
```
Comment (gadmm) [inline runtime/caml/mlvalues.h:199]: A while back I proposed mixed-size atomic accesses to solve issues with races on the header, and the answer was essentially that they are a can of worms memory-model-wise. Now the intent here seems to do a mixed-size relaxed atomic accesses. I am generally in favour of this solution as I did not understand the objections (in particular the Linux kernel does not seem to have problems with mixed-size accesses). But I am curious about  hearing objections people might have. (cc @maranget?)

The alternative is to do a volatile load of the full word followed by truncating the value.

File: runtime/major_gc.c [hunk 4]
```diff
@@ -910,7 +910,7 @@ Caml_noinline static intnat do_some_marking(struct mark_stack* stk,
     for (; me.start < scan_end; me.start++) {
       CAMLassert(budget >= 0);
 
-      value child = *me.start;
+      value child = atomic_load_relaxed_val(me.start);
```
Comment (gadmm) [inline runtime/major_gc.c]: The discussion above is a bit hard to follow because concerns are mixed with TSan considerations, but my suggestion was to change:
```c
typedef struct {
  value* start;
  value* end;
} mark_entry; /* represents fields in the span [start, end) */
```
to:
```c
typedef struct {
  volatile value * start;
  volatile value * end;
} mark_entry; /* represents fields in the span [start, end) */
```
(At this point I'd introduce `typedef volatile value * value_ptr` to denote pointers into the OCaml heap in the runtime codebase.)

In this case we avoid introducing `atomic_load_relaxed_val`.

Would this fix be enough?

**Review by gadmm — COMMENTED** [review]
Thanks for re-focusing this PR on the proposed changes independently from TSan. Here's a review without TSan in mind.

File: runtime/major_gc.c (hunk 4 above)
Comment (fabbing) [reply runtime/major_gc.c]: I also think it's the way to go for correctness if we can't make them atomic.

Just to be sure, I think I've read in your #10992 PR or in one link provided as source that having volatile qualifier may have an impact on performance (preventing compiler optimisations).
This shouldn't be the case here, until we dereference the pointer, should it? `do_some_marking` plays a lot with `me.start` and we don't want to have a negative impact on performance.

File: runtime/caml/mlvalues.h (hunk 3 above)
Comment (OlivierNicole) [reply runtime/caml/mlvalues.h:199]: Just yesterday, @damiendoligez told us some scary stories about this: in some cases of races involving mixed-size accesses, the resulting behavior is wrong non-memory-model-compliant no matter the number of fences you add, IIUC.

Comment (gadmm) [reply runtime/caml/mlvalues.h:199]: I would like to hear more about the argument. I know that 1) there are issues with sequential consistency (not relevant here), 2) the x86 specification does not specify them but they are known to work in practice and are widely relied-upon, with a few caveats. For instance on x86 you must make sure that the accesses do not cross a cacheline boundary ([source](https://lists.isocpp.org/std-proposals/2023/04/6237.php)). Since values are aligned, this constraint is satisfied here.

File: runtime/major_gc.c (hunk 4 above)
Comment (gadmm) [reply runtime/major_gc.c]: (You _could_ replace volatile with _Atomic and a relaxed load here, but use volatile for consistency with the rest of the codebase—it is better to discuss this style issue elsewhere such as #10992.)

It is the only read of this value, so there is nothing to optimize here. In addition, it is good to have in mind that the kind of optimization volatile/relaxed atomic prevents are in general not useful for code that is already carefully optimized by hand.

Note that making this change incites to add volatile to the definition of `Op_val`. This might be a thread to start unraveling, in relationship with #11426.

File: runtime/caml/mlvalues.h (hunk 3 above)
Comment (OlivierNicole) [reply runtime/caml/mlvalues.h:199]: I’m afraid I don’t know enough to offer more than this imperfect second-hand account, unfortunately.

File: runtime/caml/mlvalues.h (hunk 2 above)
Comment (OlivierNicole) [reply runtime/caml/mlvalues.h]: Ah, no. We could un-instrument `caml_initialize`, but C stubs that initialize using plain assignments or `Field` can still in principle cause false positives. (To date we have not observed such)

**Comment by fabbing** [comment]
The PR now contains all the agreed changes.
It only failed the CI because of the missing `Changes` entry, which isn't probably needed.

**Review by gadmm — COMMENTED** [review]
Thanks, briefly:
- I'd like it if there was a second opinion on this mixed-size atomic access business (@stedolan?) even if it is a very mild instance of it. The alternative is to load the full word and coerce (`Tag_val` is only expected to appear as an r-value nowadays; in addition this rids you of the endianness business).
- Please remove from the PR `atomic_load_relaxed_val` which is no longer used.
- Let's include a Changes entry anyway
- This incites generalizing the use of the `value_ptr` type alias, and in particular fix `Op_val` to correctly coerce to `value_ptr`. However, this is best done separately.

**Comment by fabbing** [comment]
> * Please remove from the PR `atomic_load_relaxed_val` which is no longer used.

Good point, I forgot to remove this function!

> * Let's include a Changes entry anyway

Done with a more suitting PR title.

**Comment by OlivierNicole** [comment]
Regarding the mixed-size `volatile` access, I agree that it would be good to have a second opinion.

File: runtime/caml/mlvalues.h (hunk 2 above)
Comment (OlivierNicole) [reply runtime/caml/mlvalues.h]: > I don't agree with your reasoning. It took a lot of discussion to decide on which style we wanted for our C code, and https://github.com/ocaml/ocaml/issues/10992 proposed to use volatile to enforce our memory model, for performance reasons. We collectively agreed on that.

I can’t find those performance reasons you mention. Could you point me to them? In the linked issue `volatile` is clearly advocated as a stop-gap measure for existing code, as @gadmm said.

Comment (gasche) [reply runtime/caml/mlvalues.h]: My understanding (but I'm not the expert in this discussion thread):
- for runtime data structures that are only used from C code, we should use C11 atomics 
- but for the OCaml heap and any other data that is accessed both from the C runtime and from the OCaml mutator, we currently use `(volatile *)` following the Linux model:
  + using C11 atomics does not provide any correctness guarantees in presence of races coming from OCaml accesses, we need to reason on the assembly level anyway
  + using `relaxed` may in general not be correct (we typically don´t want to have the reads hoisted out of loops, etc.)
  + using `consume` or `acquire` or `sequential` may be too expensive

Comment (OlivierNicole) [reply runtime/caml/mlvalues.h]: Thank you, I had missed the fact that `relaxed` may not be correct.

Comment (gadmm) [reply runtime/caml/mlvalues.h]: There are two aspects.
- Why do we use volatile instead of relaxed atomic -> because of backwards compatibility.
- Why do we use relaxed atomic or volatile instead of an acquire load (and cross our collective fingers regarding dependency ordering) -> one reason can be performance (see the bad performance of the SRA model in the memory model paper).

I think that there is no good reason to break compatibility if not for going all the way to a standards-compliant solution (i.e. acquire).

One model at #10992 advocated in favour of strict C11 compliance at the expense of backwards-compatibility (which would only be preserved for single-threaded C FFI code or code synchronized separately, e.g. with a mutex). It wants to add a `Load_field` macro that does an acquire load but is only strictly necessary for mutable fields (of non-immediate types, probably).

The other solution is to try to make sense of the current usage of dependency ordering in terms of an extension of the C11 memory model, and working together with people who would like to have some form of dependency ordering in C to add it to the standard.

The latter seems a bit more far-fetched, but from private discussions, there might be a possible path (though we probably want more than a theoretical possibility, we want people actually working on it). Regarding the former, some devs seemed to agree in private that asking people to move from `Field` to `Load_field` with no observable benefit is doomed to fail—there will always remain incorrect uses of `Field` in the wild and it might even be the majority. Though it would have the benefit for OCaml devs of shifting responsibility onto programmers.

In the meanwhile, using `volatile` is indeed a stop-gap measure, we agree it is not good, and there may be some conclusions to draw at some point from the unwillingness or inability to discuss the issue at #10992. That's why any input to continue the discussion at #10992, in particular experience from implementing the data-race detector, is welcome!

**Comment by avsm** [comment]
Triage meeting discussion summary: @damiendoligez says these accesses happen at the cache-line level these days, so a byte/word shouldn't make a difference. When a byte/word is written, we get exclusive access to the cacheline and once you have exclusive ownership then it's invalidated everyone else's cache.  The endianness question doesn't matter for the memory model, but it might affect code cleanliness.   Requests second opinion from @maranget

File: runtime/caml/mlvalues.h (hunk 3 above)
Comment (maranget) [reply runtime/caml/mlvalues.h:199]: If the question is given  (1) access the word and mask; or (2) access the byte directly, are (1) and (2)  equivalent in behaviour, I'd venture to say yes, provided all other accesses are to the complete word of to the header byte. Otherwise replacing the complete access by the header byte access would suppress some conflicts of accesses and would potentially introduced relaxed behaviours.  Of course I have no proof...

Mixed sized models for x86 and ARMv8 are studied in [Armed Cats: formal concurrency modelling at Arm](https://inria.hal.science/hal-03470858). Although it is not easy to state that the header access as a byte is "safe" from the models, I can at least say that (aligned) conflicting writes  (_i.e._ writes that have some bytes in common) are ordered, and that no observation can contradict this ordering. This property is a slight extension of the same basic property applied to non-mixed-size accesses are suggests that nothing nasty can occur when some accesses are replaced by access to the same subpart (here the header).

**Comment by Octachron** [comment]
My understanding of the discussion is that the PR has converged to a minimal safe PR that strictly improves the current state with a small change of internal API.

I have discussed with @maranget about mixed-size access, and my personal conclusion from my limited understanding was that we really want to keep our mixed-size accesses overlapped
to stay in the not-formally-proved-but-probably-safe subset of mixed size accesses.

Fortunately, this is the case if we only mix accesses on either the tag or the full word.

My proposal is thus to merge the change in 5.1 (seen as a still experimental release) before the first beta.

**Comment by gadmm** [comment]
Thank you @maranget for the detailed comment on mixed-size accesses (https://github.com/ocaml/ocaml/pull/12030#discussion_r1247830087), this addresses the last remaining question of my review.

**Comment by kayceesrk** [comment]
@fabbing @OlivierNicole can you fix the conflicts in the Changes file and ensure that all reviewers are attributed?

**Review by Octachron — APPROVED** [review]
Approving on behalf of @gadmm.

**Comment by OlivierNicole** [comment]
Sorry, I haven’t updated the reviewers yet. Doing that now…

**Comment by OlivierNicole** [comment]
Done.

**Comment by Octachron** [comment]
Cherry-picked on 5.1 as 122e20284e .

File: Changes [hunk 5]
```diff
@@ -178,6 +178,9 @@ Working version
   (Vincent Laviron, report by François Pottier, review by Nathanaëlle Courant
   and Gabriel Scherer)
 
+- #12037: Fix some data races by using volatile when necessary
```
Comment (lthls) [inline Changes:181]: This is the wrong PR number (I noticed that while rebasing one of my recent PRs). @Octachron since this has been cherry-picked to 5.1 you might want to fix it there too.
